### PR TITLE
Enable model soft schema validation for model instance documents

### PR DESCRIPTION
**Pull request by pawartur**
## Description

Add schema enforcement to Model Instance Document based on the linked 'model' stream

Use draft 2020-12 for validation: https://ajv.js.org/json-schema.html#draft-2020-12

## How Has This Been Tested?

The change has been tested with the new `glaze model-instance:create` command which is available on this branch (which shouldn't be merged, before this change is merged to prevent creating non-validated MIDs): https://github.com/ceramicstudio/js-glaze/tree/feat/implement-model-instance-create-APPS-581

- [x] Created tests for `SchemaValidation` in `stream-model-instance-handler` 
- [x] Updated unit tests for `ModelInstanceDocument`
- [x] Updated unit tests for `ModelInstanceDocumentHandler`  
- [x] Ran manual tests using the `glaze model-instance:create` command:
```
☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js did:create
✔ Created DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ with seed a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2

☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js composite:create ~/full.validation.composite.schema -k a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2 -o ~/encoded.full.validation.composite.json
ℹ Using DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ
✔ Composite was created and its encoded representation was saved in /Users/arturwdowiarski/encoded.full.validation.composite.json

☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js model-instance:create kjzl6hvfrbw6c85c5o0w01biip7ql2i1jazv8ze48wf8izrgtvic90g0u75x1fi '{}' -k a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2
ℹ Using DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ
✖ HTTP request to 'http://localhost:7007/api/v0/streams' failed with status 'Internal Server Error': {"error":"Validation Error: data must have required property 'stringProperty', data must have required property 'intProperty', data must have required property 'floatProperty'"}

☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js model-instance:create kjzl6hvfrbw6c85c5o0w01biip7ql2i1jazv8ze48wf8izrgtvic90g0u75x1fi '{"stringProperty":"a","intProperty":1,"floatProperty":1,"arrayProperty":[1],"stringArrayProperty":["a"]}' -k a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2
ℹ Using DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ
✖ HTTP request to 'http://localhost:7007/api/v0/streams' failed with status 'Internal Server Error': {"error":"Validation Error: data/intProperty must be >= 2, data/arrayProperty must NOT have fewer than 2 items, data/floatProperty must be >= 3, data/stringProperty must NOT have fewer than 3 characters, data/stringArrayProperty/0 must NOT have fewer than 2 characters"}

☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js model-instance:create kjzl6hvfrbw6c85c5o0w01biip7ql2i1jazv8ze48wf8izrgtvic90g0u75x1fi '{"stringProperty":"abcdefghi","intProperty":101,"floatProperty":110.5,"arrayProperty":[1,2,3,4,5],"stringArrayProperty":["abcdefghi"]}' -k a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2
ℹ Using DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ
✖ HTTP request to 'http://localhost:7007/api/v0/streams' failed with status 'Internal Server Error': {"error":"Validation Error: data/intProperty must be <= 100, data/arrayProperty must NOT have more than 4 items, data/floatProperty must be <= 110, data/stringProperty must NOT have more than 8 characters, data/stringArrayProperty/0 must NOT have more than 6 characters"}

☁  cli [feat/implement-model-instance-create-APPS-581] ⚡ node bin/run.js model-instance:create kjzl6hvfrbw6c85c5o0w01biip7ql2i1jazv8ze48wf8izrgtvic90g0u75x1fi '{"stringProperty":"abcdefgh","intProperty":100,"floatProperty":110,"arrayProperty":[1,2,3],"stringArrayProperty":["abcdef"]}' -k a0fe1bc2c191d02df085b6d21c84f555ff9b91bf05d8addc53314563c18ed7a2 
ℹ Using DID did:key:z6MkwUcEA5722i5G5AU9zQpG6FZSafVw3BLR4gU5vK1GDKTJ
✔ Created model instance with stream id: kjzl6kcym7w8y8rpc372wea7wfr9db72t4r2ovi78advvx11u060d9rnd4q5woy
```
Used the the `full.validation.composite.schema` file with the following contents:
```
type TestModel @model(accountRelation: LINK, description:"Test model") { 
  arrayProperty: [Int] @arrayLength(min:2, max:4)
  stringArrayProperty: [String] @length(min:2, max:6)
  stringProperty: String! @length(min:3, max:8)
  intProperty: Int! @intRange(min:2, max:100)
  floatProperty: Float! @floatRange(min:3, max:110)
}
```

While testing noticed that GraphQL SDL schemas with required array properties (e.g. `[String]!`) are not processed property when trying to create a composite (an error is thrown). This will be fixed in a separate PR.

## PR checklist

Before submitting this PR, please make sure:

- [ ] I have tagged the relevant reviewers and interested parties

## References:

Please list relevant documentation (e.g. tech specs, articles, related work etc.) relevant to this change, and note if the documentation has been updated.

### CODE REVIEW

File: packages/stream-model-instance/src/model-instance-document.ts [hunk 1]
```diff
@@ -198,7 +198,7 @@ export class ModelInstanceDocument<T = Record<string, any>> extends Stream {
    */
   private static async _makeGenesis<T>(
     signer: CeramicSigner,
-    content: T | null,
+    content: T,
```
Comment (stbrody) [inline packages/stream-model-instance/src/model-instance-document.ts:201]: :pray:

File: packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts [hunk 2]
```diff
@@ -54,13 +55,66 @@ const FAKE_CID_2 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54ad
...
 )
 
-const CONTENT0 = { myData: 0 }
-const CONTENT1 = { myData: 1 }
-const CONTENT2 = { myData: 2 }
+const FAKE_STREAM_ID_EMPTY_SCHEMA = StreamID.fromString(
+  'kjzl6hvfrbw6cbe494mf4cbhmkxbeu5soy2t5iyrs761flvrk5emxpzcjeq5sz4'
+)
+
+const CONTENT_VALID_1 = {
```
Comment (stbrody) [inline packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:67]: the changes to this test feel like they are overly complicating this test file and obfuscating the core of what is under test here.  I don't think you need to re-create all the tests that you added in `schema-utils.test.ts`, that is redundant. I think just a simple test to check that some schema validation is run at all would suffice, and then leave the schema-utils test to test the mechanics of schema validation itself.

I'd recommended just having a very simple schema that allows a single `myData` integer field (basically the same format as the original data used in the test before your change), adding a single test for a document that doesn't conform to that simple schema, and otherwise leaving this test file basically as it was

File: packages/canary-integration/src/__tests__/model-instance-document.test.ts [hunk 3]
```diff
@@ -154,91 +367,25 @@ describe('ModelInstanceDocument API http-client tests', () => {
...
-    const doc = await ModelInstanceDocument.create(ceramic, CONTENT0, METADATA)
+    const doc = await ModelInstanceDocument.create(ceramic, CONTENT_VALID_1, midMetadata)
     await expect(isPinned(ceramic, doc.id)).resolves.toBeTruthy()
-    await doc.replace(CONTENT1, { pin: false })
+    await doc.replace(CONTENT_VALID_2, { pin: false })
     await expect(isPinned(ceramic, doc.id)).resolves.toBeFalsy()
   })
 })
-
-describe('ModelInstanceDocument API multi-node tests', () => {
```
Comment (stbrody) [inline packages/canary-integration/src/__tests__/model-instance-document.test.ts:180]: why were these tests deleted?

File: packages/canary-integration/src/__tests__/model-instance-document.test.ts [hunk 4]
```diff
@@ -2,36 +2,139 @@ import { jest } from '@jest/globals'
...
 import { createCeramic } from '../create-ceramic.js'
 import { anchorUpdate } from '@ceramicnetwork/core/lib/state-management/__tests__/anchor-update'
 import { Ceramic } from '@ceramicnetwork/core'
 import { CeramicDaemon, DaemonConfig } from '@ceramicnetwork/cli'
 import { CeramicClient } from '@ceramicnetwork/http-client'
 import { StreamID } from '@ceramicnetwork/streamid'
 import first from 'it-first'
+import { Model, ModelAccountRelation, ModelDefinition } from '@ceramicnetwork/stream-model'
+
+const CONTENT_VALID_1 = {
```
Comment (stbrody) [inline packages/canary-integration/src/__tests__/model-instance-document.test.ts:15]: same comment here as I made below on `model-instance-document-handler.test.ts`.  Adding too much into one test obfuscates the core of what we're trying to get out of that test.  I think the tests at each layer should focus on testing what that layer is adding, while relying on tests of lower layers to test that lower-level functionality.  I think we can mostly put this test file back to the way it was and just have a simple test that schema enforcement is happening, while relying on the lower-level tests to cover all the edge cases of schema validation.

**Review by stbrody — COMMENTED** [review]
looks good, some comments on testing strategy

File: packages/canary-integration/src/__tests__/model-instance-document.test.ts [hunk 5]
```diff
@@ -52,33 +74,50 @@ describe('ModelInstanceDocument API http-client tests', () => {
     await ipfs.stop()
   })
 
-  test('Create valid doc', async () => {
-    const doc = await ModelInstanceDocument.create(ceramic, CONTENT0, METADATA)
+  test('verifies the content against model schema', async () => {
```
Comment (stbrody) [inline packages/canary-integration/src/__tests__/model-instance-document.test.ts:77]: might be good to have a second test for updating the doc to an invalid schema - so we have both the genesis commit and applying a signed commit paths tested

File: packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts [hunk 6]
```diff
@@ -326,10 +423,19 @@ describe('ModelInstanceDocumentHandler', () => {
   })
 
   it('makes signed commit correctly', async () => {
+    const content1 = {
```
Comment (stbrody) [inline packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:426]: can stick with the CONTENTX constants defined above, no need to introduce new content vars in the test

Comment (stbrody) [reply packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:426]: In fact, I think all changes to this test case can be reverted?

File: packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts [hunk 7]
```diff
@@ -455,88 +561,10 @@ describe('ModelInstanceDocumentHandler', () => {
...
-      commit: sPayload,
-      envelope: signedCommit.jws,
-    }
-    state = await handler.applyCommit(signedCommitData, context, state)
-    delete state.metadata.unique
-    delete state.next.metadata.unique
-    expect(state).toMatchSnapshot()
+  test('verifies the content against model schema', async () => {
+    await expect(
+      applyGenesisCommit(handler, {}, context)
```
Comment (stbrody) [inline packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:566]: if the applyGenesisCommit function is only used in one test, I'd just inline it

Comment (stbrody) [reply packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:566]: Would also be good to add a second test for applying a signed commit that doesn't conform to the schema, and inline and remove the applySignedCommitToValidMID function (which I believe is currently unused)

**Review by stbrody — APPROVED** [review]
lgtm mod further minimizing the changes to the existing tests, and adding test cases for the higher-level tests for updating existing streams

File: packages/stream-model-instance-handler/package.json [hunk 8]
```diff
@@ -42,12 +42,15 @@
...
   "devDependencies": {
     "@ceramicnetwork/3id-did-resolver": "^2.1.1",
+    "@ceramicnetwork/stream-model": "^0.2.0-rc.2",
     "@ipld/dag-cbor": "^7.0.0",
     "@stablelib/sha256": "^1.0.1",
     "@types/lodash.clonedeep": "^4.5.6",
     "ceramic-cacao": "^1.1.1",
     "did-resolver": "^3.1.5",
     "dids": "^3.2.0",
+
```
Comment (stbrody) [inline packages/stream-model-instance-handler/package.json:52]: rm extra newlines

**Review by stbrody — APPROVED** [review]
looks great, nice work!
